src/main/java/kitchenpos/eatinorders/tobe/domain/EatInOrderService.java

```diff
@@ -0,0 +1,16 @@
+package kitchenpos.eatinorders.tobe.domain;
+
+import kitchenpos.core.annotation.CommandDomainService;
+import kitchenpos.core.constant.ExceptionMessages;
+
+@CommandDomainService
+public class EatInOrderService {
+
+	public EatInOrder order(OrderLineItems orderLineItems, OrderTable orderTable) {
```

Comment:
```suggestion
        private final OrderTableRepository orderTableRepository;
      
	public EatInOrder order(OrderLineItems orderLineItems, UUID orderTableId) {
	    orderTableRepository.findById(orderTableId)... 
```

도메인 서비스에 엔티티를 전달하는것보다 ID만 전달하여 서비스 레이어에서 서로 다른 Aggregate간 참조를 없애는것을 권장드려요. 
위와 같이 구현한다면 Service Layer 에서의 결합도가 낮아지고 결합이 발생하는 부분은 오직 도메인서비스만 남기 때문에 추후 OrderTable 이 별도의 마이크로 서비스로 분리된다하더라도 repository 참조를 api client로 변경시 유연하게 처리할수 있습니다 😄 

이와 같이 구현하는것에 대해서 정문님의 생각이 궁금합니다 😄

Reply:
저는 에릭 에반스의 책을 읽으면서 도메인 서비스를 이해했는데 책(p.109)을 조금 인용해보자면 다음과 같습니다.
> 도메인의 개념 가운데 객체로는 모델에 어울리지 않는 것이 있다. 필요한 도메인 기능을 Entity나 Value에서 억지로 맡게 하면 모델에 기반을 둔 객체의 정의가 왜곡되거나, 또는 무의미하고 인위적으로 만들어진 객체가 추가될 것이다.  
도메인의 중대한 프로세스나 변환 과정이 ENTITY나 VALUE OBJECT의 고유한 책임이 아니라면 연산을 SERVICE로 선언되는 독립 인터페이스로 모델에 추가하라. 모델의 언어라는 측면에서 인터페이스를 정의하고 연산의 이름을 UBIQUITOUS LANGUAGE의 일부가 되게끔 구성하라. SERVICE는 상태를 갖지 않게 만들어라.

제가 이해한 바로는 도메인 서비스는 도메인 모델의 일부이나 행위에 초점이 맞춰진 것이라 할 수 있습니다. 따라서 Repository에 대한 의존성을 가질 수 없고 Repository가 도메인에 누락된 기능을 호출하기 위해 Domain Service에 대한 의존성을 가져야 합니다. 이전 질문에서 왜 도메인 서비스에서 Transactional로 구분하는지 여쭤본 것도 위와 같은 이유입니다. Domain Service는 Layered Architecture의 Service와는 분명 다른 것이기 때문에 여기서 다른 Aggregate에 대한 조회가 일어나서는 안 된다고 생각합니다.

Reply:
남겨주신 코멘트 잘 보았습니다. 확고한 기준을 가지고 구현에 녹여주셔서 저도 많이 배우고 있습니다 감사합니다 ㅎㅎ 🙏 

저는 에릭 에반스의 책을 읽으며 이해가 되지 않던 부분을 반 버논의 도메인 주도 설계 구현이라는 책을 보며 개념을 잡곤 하였습니다. 
반 버논의 책 중 도메인 서비스 챕터에 보면 아래와 같은 글이 있습니다. 책 내용과는 조금 다르게 다듬어 보았어요 😄 

```
애플리케이션 서비스 클라이언트에서 비즈니스 요구사항에 관한 세부사항을 도메인 서비스로 밀어낸다. 서비스가 필요에 따라 어떤 수의 도메인 객체든 사용할 수 있다. 클라이언트(어플리케이션 서비스)는 이러한 세부사항에 관해선 전혀 알 필요가 없다. 용어의 일부는 모델에, 일부는 클라이언트에 표현하는 대신 ID 관리 도메인을 모델링하는 소프트웨어로 올바른 용어를 표현했기 때문에 해당 컨텍스트의 유비쿼터스 언어도 만족된다. 
```

여기서 나온 이야기를 채용하자면 도메인 서비스는 비즈니스 로직을 만족하기 위해서 어떤 수의 도메인 객체든 사용할수 있고, 어플리케이션 서비스는 세부사항을 알필요가 없는것을 알수 있어요. ID 관리 도메인이라는것은 해당 프로젝트에서 Entity의 식별자로 이해해주시면 됩니다. 
ID 관리 도메인을 활용하여 클라이언트에 표현했기 때문에 해당 컨텍스트의 유비쿼터스 언어가 만족된다고 했는데, Service Layer에서 다른 컨텍스트의 Repository에 대한 의존을 가져 조회하는 경우 해당 컨텍스트의 유비쿼터스 언어가 만족되지 않는다고 생각해요. (다른 컨텍스트의 도메인이 Service Layer에 침범하기 때문이라고 생각합니다.) 
해당 책의 예제도 도메인 서비스에서 Repository를 활용하여 다른 컨텍스트의 도메인을 조회하고 있습니다. 도메인 서비스만이 여러개의 도메인을 참조할수 있기 때문이죠.

이와 같은 구현 방법을 본다면 Repository 가 도메인의 누락된 기능을 호출하기 위해 Domain Service에 대한 의존을 가지기 보다는 도메인의 요구사항을 구현하기 위해 도메인 서비스는 다른 컨텍스트의 Repository에 대한 의존을 가질수 있다. 라고 생각하게 되었습니다. 
여기서의 Repository는 현재 구현해주신 JPA Repository 가 될수 있고, 마이크로 서비스의 경우 API Client 가 될수 있다고 생각해요.

이와같은 내용은 제 생각이기 때문에 정답은 아닐거에요. ㅎㅎ 
저희 팀에서도 DDD를 접목시킬때 도메인 서비스에 대해서 많은 논의와 구현방법에 대해서 고민하였는데, 정문님의 의견을 한번더 요청드리고 싶습니다 🙏

Reply:
저도 동철님과 도메인 주도 설계에 대해 깊게 얘기해볼 수 있어 의미 있는 시간인 것 같습니다.

결국엔 Domain Service를 무엇으로 정의할 것이냐에 따라 설계가 달라지는 것 같습니다. 저는 둘 이상의 엔티티와 관련된 도메인 로직을 실행하는 주체로 봤습니다. 하나의 엔티티에 억지로 위치 시킨다면 캡슐화가 깨지는 것이라는 설명이 있는데 그렇다면 도메인 서비스는 행위에 집중한 도메인 모델이 아닐까 하는 생각입니다. 애플리케이션 서비스에서 비지니스 로직을 밀어내기 위해 사용하는 건 맞지만 그로 인해 도메인 모델이 영속 계층에 대한 의존성을 가져야 한다는 걸 의미하진 않는다 생각합니다. 일반적으로 도메인 계층은 어디에도 의존하지 않게 설계하니 Domain Service 또한 마찬가지라 생각합니다. 아무리 Repository 인터페이스에 의존하게 만들고 DI로 런타임에 구현체를 주입하더라도 도메인에서 영속화와 관련된 의존성을 갖는 것 자체가 어색하게 느껴졌습니다. 

말씀주신 내용에서 `다른 컨텍스트의 도메인이 Service Layer에 침범하기 때문이라고 생각합니다.` 부분은 Application Service 네이밍을 OrderService와 같이 도메인 명에 Service만 붙여 만들지 않고 의미 있는 기능 단위로 만들면 된다고 생각합니다. 그렇게 한다면 주문 기능을 수행하기 위해서는 매장 주문과 주문 테이블이 필요하겠구나 라고 쉽게 파악할 수 있을 것 같습니다. 매장 주문과 주문 테이블은 같은 매장 주문 컨텍스트에 위치해 있기도 하구요. 

결론으로 Domain Service의 책임에 영속화가 속할까? 라는 의문에 저는 아니라고 판단했습니다. 지극히 제 개인적인 생각이고 정답은 없기에 팀 합의로 얼마든지 바꿀 수 있는 내용일 것 같습니다.

Reply:
마지막 말씀대로 팀내에서 어떻게 정의하냐에 따라서 결정될것 같아요 😄 
정문님과 의견을 주고 받으면서 도메인 서비스에 대해서 더 많은 고민을 하게된 기회였던것 같습니다 !